File: linked_list/lru_cache.js
function Node(key, val, next, prev) {
  this.key = (key===undefined) ? 0 : key;
  this.val = (val===undefined ? 0 : val);
  this.next = (next===undefined ? null : next);
  this.prev = (prev===undefined ? null : prev);
}

var LRUCache = function(capacity) {
  this.capacity = capacity;
  this.size = 0;
  this.cache = {};
  this.left = new Node(0,0);
  this.right = new Node(0,0);
  this.left.next = this.right;
  this.right.prev = this.left;
};

LRUCache.prototype.get = function(key) {
  if(key in this.cache) {
    this.remove(this.cache[key]);
    this.insert(this.cache[key]);
    return this.cache[key].val;
  }
  return -1;
};

/** 
* @param {number} key 
* @param {number} value
* @return {void}
*/
LRUCache.prototype.put = function(key, value) {
  if(key in this.cache) {
    this.remove(this.cache[key]);
  }
  this.cache[key] = new Node(key, value);
  this.insert(this.cache[key]);

  if(this.capacity < this.size) {
    let lru = this.left.next;
    this.remove(lru);
    delete this.cache[lru.key];
  }
};

LRUCache.prototype.remove = function(node) {
  if(!(this.size === 0)) {
    let nextNode = node.next;
    let prevNode = node.prev;
    prevNode.next = nextNode;
    nextNode.prev = prevNode;
    this.size--;
  }
}

LRUCache.prototype.insert = function(node) {
  let prevNode = this.right.prev;
  prevNode.next = node;
  node.prev = prevNode;
  node.next = this.right;
  this.right.prev = node;
  this.size++;
}


var obj = new LRUCache(2);
console.log(obj.get(2));
obj.put(1,1);
obj.put(2,2);
console.log(obj.get(1));
obj.put(3,3);
console.log(obj.get(2));
obj.put(4,4);
console.log(obj.get(3));

File: old/linked_list/lru_cache.py
class Node:
    def __init__(self, key=0, val=0, prev=None, next=None):
        self.key = key
        self.val = val
        self.prev = prev
        self.next = next

class LRUCache:

    def __init__(self, capacity: int):
        self.cap = capacity
        self.size = 0
        self.cache = {}
        self.left = Node(0, 0)
        self.right = Node(0, 0)
        self.left.next = self.right
        self.right.prev = self.left

    def remove(self, node):
        if not self.size:
            prev = node.prev
            next = node.next
            prev.next = next
            next.prev = prev
            self.size -= 1
    
    def insert(self, node):
        prev = self.right.prev
        prev.next = node
        node.next = self.right
        node.prev = prev
        self.right.prev = node
        self.size += 1

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self.remove(node)
            self.insert(node)
            return node.val
        return -1
        

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.remove(self.cache[key])
        self.cache[key] = Node(key, value)
        self.insert(self.cache[key])

        if self.cap < self.size:
            left = self.left.next
            self.remove(left)
            del self.cache[left.key]


# Your LRUCache object will be instantiated and called as such:
obj = LRUCache(1)
param_1 = obj.get(1)
obj.put(2,3)
obj.put(1,3)